sw/device/boot_rom/rom_link.ld

```diff
@@ -2,106 +2,140 @@
+
 _chip_info_size = 128;
 _chip_info_start = ORIGIN(rom) + LENGTH(rom) - _chip_info_size;
 
-/* We have to align each sector to word boundaries as our current s19->slm
- * conversion scripts are not able to handle non-word aligned sections. */
-
-SECTIONS
-{
-    .vectors :
-    {
-        . = ALIGN(4);
-        KEEP(*(.vectors))
-    } > rom
-
-    .text : {
-        . = ALIGN(4);
-        _stext = .;
-        *(.text)
-        *(.text.unlikely)
-        _etext  =  .;
-        __CTOR_LIST__ = .;
-        LONG((__CTOR_END__ - __CTOR_LIST__) / 4 - 2)
-        *(.ctors)
-        LONG(0)
-        __CTOR_END__ = .;
-        __DTOR_LIST__ = .;
-        LONG((__DTOR_END__ - __DTOR_LIST__) / 4 - 2)
-        *(.dtors)
-        LONG(0)
-        __DTOR_END__ = .;
-        *(.lit)
-        *(.shdata)
-        _endtext = .;
-    }  > rom
-
-    .rodata : {
-        . = ALIGN(4);
-        *(.rodata);
-        *(.rodata.*)
-    } > rom
-
-    .shbss :
-    {
-        . = ALIGN(4);
-        *(.shbss)
-    } > rom
-
-    .data : {
-        . = ALIGN(4);
-        sdata  =  .;
-        _sdata  =  .;
-        *(.data);
-        *(.data.*)
-        edata  =  .;
-        _edata  =  .;
-    } > rom
-
-    .chip_info _chip_info_start : {
-        . = ALIGN(4);
-        *(.chip_info)
-    } > rom
-
-    .bss :
-    {
-        . = ALIGN(4);
-        _bss_start = .;
-        *(.bss)
-        *(.bss.*)
-        *(.sbss)
-        *(.sbss.*)
-        *(COMMON)
-        _bss_end = .;
-    } > ram
-
-    .stab  0 (NOLOAD) :
-    {
-        [ .stab ]
-    }
-
-    .stabstr  0 (NOLOAD) :
-    {
-        [ .stabstr ]
-    }
+/**
+ * NOTE: We have to align each sector to word boundaries as our current s19->slm
+ * conversion scripts are not able to handle non-word aligned sections.
+ */
+SECTIONS {
+  /**
+   * Ibex interrupt vector. See irq_vector.S for more information. 
+   *
+   * This has to be set up at the boot address, so that execution jumps to the
+   * reset handler correctly.
+   */
+  .vectors _boot_address : {
+    . = ALIGN(4);
+    KEEP(*(.vectors))
+    *(.vectors)
+  } > rom
+
+  /**
+   * C runtime (CRT) segment, containing program initialization code.
+   */
+  .crt : {
+    . = ALIGN(4);
+    KEEP(*(.crt))
+    *(.crt)
+  } > rom
+
+  /**
+   * Standard text segment, containing program code.
+   */
+  .text : {
+    . = ALIGN(4);
+    *(.text)
+    *(.text.unlikely)
+  } > rom
+
+  /**
+   * Read-only data segment, containing all large compile-time constants, like
+   * strings.
+   */
+  .rodata : {
+    . = ALIGN(4);
+    *(.rodata)
+    *(.rodata.*)
+  } > rom
+
+  /**
+   * "Intitial data" segment, the initial values of the mutable data segment 
+   * initialized at runtime.
+   */ 
+  .idata : {
+    . = ALIGN(4);
+    _data_init_start = .;
+  } > rom
+
+  /** 
+   * Standard mutable data segment, at the bottom of RAM. This will be
+   * initialized from the .idata segment at runtime.
+   */ 
+  .data ORIGIN(ram): AT(_data_init_start) {
```

Comment:
I think using `AT(_data_init_start)` will force the LMA to be `_data_init_start`, but will not allocate space in the `rom` region, which could let the linker place things over top of them if they follow `.idata`.  I think you need to use `AT>` to make sure there's allocated space, but @lenary could probably tell me I'm wrong here.

Reply:
This is cribbed from the flash executable linker script; I honestly don't know too much about making LMA/VMA do the right thing.

Reply:
I am not familiar enough with linker scripts to know either. I can crib up on them tomorrow if need be.

Reply:
Having read up on `AT` vs `AT>`, only the latter mentions "next free address in the region" in the LD docs. I think that using `AT>` will make it much easier to add more segments to be loaded into rom, rather than needing to add another symbol for a second `AT(…)`. 

That said, I'm not sure how we find the LMA bounds of a section using `AT>` for doing loading in the CRT (where you copy the section from its LMA address to its VMA address).

Reply:
We could also just not bother with LMA/VMA lolz, and just do something like
```
.idata : { *(.data) } > rom
.data ORIGIN(ram):  { *(.data) } > ram
```
What do you think?

Reply:
This sounds like a bad idea, as you'd get duplicate symbol definitions, surely?

Let's use `AT(…)` for the moment, and if we decide we need `AT>`/`LOADADDR`, then we can move to that.